### DISCUSSION TITLE
How to install clangd on Ubuntu (18.04)?

### DISCUSSION THREAD

**Discussion by hrcHarry** [clangd/clangd#1687] [Q&A] [answered]
Hi. Sorry if this is a basic issue.

After downloading [clangd-linux-16.0.2.zip](https://github.com/clangd/clangd/releases/download/16.0.2/clangd-linux-16.0.2.zip) on [https://github.com/clangd/clangd/releases/tag/16.0.2](https://github.com/clangd/clangd/releases/tag/16.0.2), I have no idea how to install it on Ubuntu (18.04). What instructions should I do for installing?

By the way, what is [clangd_indexing_tools](https://github.com/clangd/clangd/releases/tag/16.0.2) used for? (Website [https://clangd.llvm.org/installation.html#installing-clangd](https://clangd.llvm.org/installation.html#installing-clangd) only shows the information about clangd). When will we use it?

Thanks.

**Comment by HighCommander4** [comment, marked as answer]
> After downloading [clangd-linux-16.0.2.zip](https://github.com/clangd/clangd/releases/download/16.0.2/clangd-linux-16.0.2.zip) on https://github.com/clangd/clangd/releases/tag/16.0.2, I have no idea how to install it on Ubuntu (18.04). What instructions should I do for installing?

Clangd is a language server meant to be used together with a client (editor that supports the Language Server Protocol).

So the general installation steps would be:
 1. Unzip
 2. Configure your editor to point it to the location of the extracted clangd binary

The way to do (2) is specific to the editor. For example, for VSCode, you would set `clangd.path` in VSCode settings to the path of the clangd binary. See https://clangd.llvm.org/installation.html#editor-plugins for more details. 

> By the way, what is [clangd_indexing_tools](https://github.com/clangd/clangd/releases/tag/16.0.2) used for? (Website https://clangd.llvm.org/installation.html#installing-clangd only shows the information about clangd). When will we use it?

The indexing tools are only applicable if using [remote indexing](https://clangd.llvm.org/guides/remote-index).

**Reply by i-ky** [reply]
As an alternative to (2) one can also make sure that `clangd` is in `PATH` (I do it with `cd /usr/local/bin/ && sudo ln -s /path/to/clangd`), this way a default editor configuration most likely will work out of the box.

**Reply by hrcHarry** [reply]
Thanks @HighCommander4 and @i-ky Both of you are awesome and deserve a thumb up! I do would like an editor (in my case, Neovim) _attach_ clangd.

And @i-ky Just for sure, and take [clangd-linux-16.0.2.zip](https://github.com/clangd/clangd/releases/download/16.0.2/clangd-linux-16.0.2.zip) for example: Does `/path/to/clangd` mean pointing to the `clangd` file (`clangd_16.0.2/bin/clangd`) or the whole directory that [clangd-linux-16.0.2.zip](https://github.com/clangd/clangd/releases/download/16.0.2/clangd-linux-16.0.2.zip) is unziped?
![](https://upload.cc/i1/2023/07/01/FYmODE.png)

I made `ln -s` both 😅 and `echo $PATH` shows none of them. Am I doing wrong at some step?

(Additional note: I use WSL2)

**Comment by i-ky** [comment]
> Does `/path/to/clangd` mean pointing to the `clangd` file (`clangd_16.0.2/bin/clangd`) or the whole directory that [clangd-linux-16.0.2.zip](https://github.com/clangd/clangd/releases/download/16.0.2/clangd-linux-16.0.2.zip) is unziped?

It should be the path to `clangd` executable file, in this case `path/to/clangd-linux-16.0.2/clangd_16.0.2/bin/clangd`. It should look like so (I build `clangd` from source myself, but you get the gist):
```
$ ls -l /usr/local/bin
...
lrwxrwxrwx 1 root root    47 Jan 26 07:48 clangd -> /home/i-ky/GitHub/llvm-project/build/bin/clangd
...
```

> I made `ln -s` both :sweat_smile: and `echo $PATH` shows none of them. Am I doing wrong at some step?

`ln` does not change the `PATH`, it is not needed. `/usr/local/bin/` is one of the directories that usually are in the `PATH` by default. You can check if everything is OK by typing
```
clangd --version
```
into your terminal. If `clangd` is in the `PATH` (or, more precisely, in one of the directories listed in `PATH`), then you should see something like
```
clangd version 17.0.0 (https://github.com/llvm/llvm-project 5344d8e10bb7d8672d4bfae8adb010465470d51b)
Features: linux+grpc
Platform: x86_64-unknown-linux-gnu
```
If instead your shell complains that there is no `clangd` to run, there is a problem.

> (Additional note: I use WSL2)

It should be no different on WSL.

**Reply by hrcHarry** [reply]
Thank you so much! How kind of you giving such a thorough explanation to make sure I can get the gist. 
You help me a lot! ❤👍
